doc/Addendum-NX-GZIP-for-PowerVM.txt

```diff
@@ -0,0 +1,317 @@
+PowerVM also adds a direct access user-space interface to nx-gzip
+eliminating the kernel and hypervisor from the software path.
+
+This document is an addendum to the nx-gzip user manual [UM] and
+describes nx-gzip access methods on PowerVM based systems. While,
+PowerVM interface is the same for both Linux and AIX, this document is
+intended for Linux developers.  For AIX, readers may refer to the
+documentation here [AIX1, AIX2].
+
+On PowerVM, the copy/paste instruction pair and the VAS window
+mechanism and data structures used in PowerNV will work identically.
+Small changes may be required in the kernel driver to open the VAS
+window through PowerVM HCALLs (see the Linux kernel implementation here
+[P1,P2].) Additionally, PowerVM implements some admission and QoS
+policies for sharing nx-gzip resources among multiple processes and
+LPARs.
+
+H_ALLOCATE_VAS_WINDOW, H_DEALLOCATE_VAS_WINDOW, H_MODIFY_VAS_WINDOW
+hcalls are described and example usage are found in the Linux kernel
+patch series detailed here [P1, P2]
+
+
+Accessing nx-gzip on PowerVM
+----------------------------
+
+On PowerVM, an nx-gzip enabled operating system such as Linux:
+
+- Must get all VAS capabilities using the H_QUERY_VAS_CAPABILITIES
+  HCALL available in PowerVM.  The capabilities tell the operating
+  system which features and credit types are available such as
+  Default and Quality of Service (QoS). The HCALL also gives specific
+  capabilities for each credit type: Maximum VAS window credits,
+  Maximum LPAR credits, Target credits in that partition (varies from
+  maximum LPAR credits based DLPAR operation), whether the interface
+  supports user mode COPY/PASTE.
+
+- Must register LPAR VAS operations such as open window, get paste
+  address and close window with the current VAS user-space API.
+
+- Must use H_ALLOCATE_VAS_WINDOW HCALL to open a VAS window and
+  H_MODIFY_VAS_WINDOW HCALL to setup the window with LPAR PID
+
+- Must mmap the H_ALLOCATE_VAS_WINDOW HCALL returned address to paste
+  CRB commands to
+
+- Must close a VAS window with H_DEALLOCATE_VAS_WINDOW HCALL
+
+For nx-gzip enablement, the user-space application:
+
+- Must get nx-gzip capabilities from the operating
+  system. Capabilities include Maximum buffer length for a single GZIP
+  request and recommended minimum compression / decompression lengths.
+
+- Must register with VAS to enable user-space VAS API
+
+
+Main PowerVM differences from PowerNV
+-------------------------------------
+
+- Each VAS window will be configured with a number of credits limiting
+  the number of nx-gzip requests that can be simultanenously sent
+  through window, administrator configurable. On PowerNV, 1024 credits
+  are configured per window.  Whereas on PowerVM, by default 1 credit
+  per window are allowed
+
+- PowerVM introduces 2 different types of credits called the Default
+  and QoS credits Default uses the normal priority nx-gzip queue. QoS

```

Comment:
Missing a period (or colon) after QoS credits.

Reply:
Thank you